extensions/wikia/PortableInfobox/services/PortableInfoboxDataService.class.php

```diff
@@ -175,13 +200,52 @@ protected function load() {
 		$id = $this->title->getArticleID();
 		if ( $id ) {
 			return WikiaDataAccess::cache( $this->cachekey, WikiaResponse::CACHE_STANDARD, function () use ( $id ) {
-				return json_decode( $this->propsProxy->get( $id, self::INFOBOXES_PROPERTY_NAME ), true );
+				return $this->reparseIfNeeded(
+					json_decode( $this->propsProxy->get( $id, self::INFOBOXES_PROPERTY_NAME ), true )
+				);
 			} );
 		}
 
 		return [ ];
 	}
 
+	/**
+	 * If PageProps doesn't have infoboxes' metadata or has an old version then reparse the page and store fresh data
+	 *
+	 * @param $infoboxes
+	 *
+	 * @return array
+	 */
+	protected function reparseIfNeeded( $infoboxes ) {
+		if ( is_null( $infoboxes ) ) {
+			return $this->reparse();
+		}
+
+		foreach ( $infoboxes as $infobox ) {
+			if (
+				empty( $infobox ) ||
+				$infobox['parser_tag_version'] !== PortableInfoboxParserTagController::PARSER_TAG_VERSION
+			) {
+				return $this->reparse();
+			}
+		}
+
+		return $infoboxes;
+	}
+
+	protected function reparse() {
```

Comment:
I'd propose to extract this method together with article content fetching to separate class. That way we can reuse it without decoupling `DataService` and `TemplateHelper`. You could also move everything into `TemplateHelper` and rename it to be a more appropriate to it's functionality.